File: unit-testing-workshop/src/com/ihsinformatics/unittesting/SemVerUtil.java
/**
 * 
 */
package com.ihsinformatics.unittesting;

import java.util.stream.Stream;

/**
 * @author owais.hussain@ihsinformatics.com
 *
 */
public class SemVerUtil {

	private static String VALID_PREALPHA_REGEX = "^0\\.0\\.[0-9]+";
	private static String VALID_ALPHA_REGEX = "^0\\.[0-9]+\\.[0-9]+\\-(alpha)";
	private static String VALID_BETA_REGEX = "^0\\.[0-9]+\\.[0-9]+\\-(beta)";
	private static String VALID_RELEASE_REGEX = "^[0-9]+\\.[0-9]+\\.[0-9]+";
	private static String VERSION_SPLITTER = "\\.";
	private static String EXTENSION_SEPARATOR = "-";

	private int major;
	private int minor;
	private int micro;
	private String extension;

	/**
	 * Pass a valid version in the string according to Semantic versioning scheme
	 * 
	 * @param version
	 */
	public SemVerUtil(String version) {
		this.parseVersion(version.toLowerCase());
	}

	/**
	 * Parse version in a string
	 * 
	 * @param version
	 */
	protected void parseVersion(String version) {
		if (version == null) {
			throw new NullPointerException("Input cannot be null");
		}
		String[] parts;
		if (version.matches(VALID_PREALPHA_REGEX)) {
			parts = version.split(VERSION_SPLITTER);
			setMajor(0);
			setMinor(0);
			setMicro(Integer.parseInt(parts[2]));
			extension = null;
		} else if (version.matches(VALID_ALPHA_REGEX)) {
			extension = "alpha";
			version = version.substring(0, version.lastIndexOf(EXTENSION_SEPARATOR));
			parts = version.split(VERSION_SPLITTER);
			setMajor(0);
			setMinor(Integer.parseInt(parts[1]));
			setMicro(Integer.parseInt(parts[2]));
		} else if (version.matches(VALID_BETA_REGEX)) {
			extension = "beta";
			version = version.substring(0, version.lastIndexOf(EXTENSION_SEPARATOR));
			parts = version.split(VERSION_SPLITTER);
			setMajor(0);
			setMinor(Integer.parseInt(parts[1]));
			setMicro(Integer.parseInt(parts[2]));
		} else if (version.matches(VALID_RELEASE_REGEX)) {
			extension = null;
			parts = version.split(VERSION_SPLITTER);
			setMajor(Integer.parseInt(parts[0]));
			setMinor(Integer.parseInt(parts[1]));
			setMicro(Integer.parseInt(parts[2]));
		} else {
			throw new NumberFormatException("Invalid input!");
		}
	}

	/**
	 * @return the major
	 */
	public int getMajor() {
		return major;
	}

	/**
	 * @param major the major to set
	 */
	public void setMajor(int major) {
		this.major = major;
	}

	/**
	 * @return the minor
	 */
	public int getMinor() {
		return minor;
	}

	/**
	 * @param minor the minor to set
	 */
	public void setMinor(int minor) {
		this.minor = minor;
	}

	/**
	 * @return the micro
	 */
	public int getMicro() {
		return micro;
	}

	/**
	 * @param micro the micro to set
	 */
	public void setMicro(int micro) {
		this.micro = micro;
	}

	/**
	 * Checks whether current version is pre-alpha or not
	 * 
	 * @return
	 */
	public boolean isPreAlpha() {
		return (getMajor() == 0 && getMinor() == 0 && extension == null);
	}

	/**
	 * Checks whether current version is alpha or not
	 * 
	 * @return
	 */
	public boolean isAlpha() {
		if (extension == null) {
			return false;
		}
		return extension.equals("alpha");
	}

	/**
	 * Checks whether current version is alpha or not
	 * 
	 * @return
	 */
	public boolean isBeta() {
		if (extension == null) {
			return false;
		}
		return extension.equals("beta");
	}

	/**
	 * Checks whether current version is a production release or not
	 * 
	 * @return
	 */
	public boolean isRelease() {
		return !(isPreAlpha() || isAlpha() || isBeta()) && extension == null;
	}

	/**
	 * Compares two versions and: returns 0 if both are same; 1 if newer is ahead of
	 * older; -1 if older is ahead of newer
	 * 
	 * @param older
	 * @param newer
	 * @return
	 */
	public static int compareVersions(SemVerUtil older, SemVerUtil newer) {
		if (older.toString().equals(newer.toString())) {
			return 0;
		}
		if (older.isPreAlpha() && !newer.isPreAlpha()) {
			return 1;
		}
		if (older.isAlpha() && (newer.isBeta() || newer.isRelease())) {
			return 1;
		}
		if (older.isBeta() && newer.isRelease()) {
			return 1;
		}
		boolean allFlags = older.isPreAlpha() == newer.isPreAlpha() 
				|| older.isAlpha() == newer.isAlpha()
				|| older.isBeta() == newer.isBeta()
				|| older.isRelease() == newer.isRelease();
		if (allFlags) {
			// If both are release, then compare major, minor and micro
			if (older.getMajor() < newer.getMajor()) {
				return 1;
			}
			if (older.getMajor() == newer.getMajor() && older.getMinor() < newer.getMinor()) {
				return 1;
			}
			if (older.getMajor() == newer.getMajor() && older.getMinor() == newer.getMinor()
					&& older.getMicro() < newer.getMicro()) {
				return 1;
			}
		}
		return -1;
	}

	/**
	 * @param semVerUtil
	 * @return
	 */
	public boolean isBackwardsCompatibleWith(SemVerUtil semVerUtil) {
		return getMajor() == semVerUtil.getMajor() && minor == semVerUtil.getMinor();
	}

	@Override
	public String toString() {
		String string = getMajor() + "." + getMinor() + "." + getMicro() + (extension == null ? "" : "-" + extension);
		return string;
	}
}

File: unit-testing-workshop/src/python/sem_ver_util.py
from __future__ import annotations

import re
from typing import Optional


_VALID_PREALPHA_REGEX = r"^0\.0\.[0-9]+$"
_VALID_ALPHA_REGEX = r"^0\.[0-9]+\.[0-9]+\-(alpha)$"
_VALID_BETA_REGEX = r"^0\.[0-9]+\.[0-9]+\-(beta)$"
_VALID_RELEASE_REGEX = r"^[0-9]+\.[0-9]+\.[0-9]+$"
_VERSION_SPLITTER = "."
_EXTENSION_SEPARATOR = "-"


class SemVerUtil:
    """Parsed semantic-version triple with optional ``alpha``/``beta`` extension."""

    def __init__(self, version: str) -> None:
        if version is None:
            raise TypeError("Input cannot be null")
        self._major = 0
        self._minor = 0
        self._micro = 0
        self._extension: Optional[str] = None
        self._parse_version(version.lower())

    def _parse_version(self, version: str) -> None:
        if re.match(_VALID_PREALPHA_REGEX, version):
            parts = version.split(_VERSION_SPLITTER)
            self.set_major(0)
            self.set_minor(0)
            self.set_micro(int(parts[2]))
            self._extension = None
        elif re.match(_VALID_ALPHA_REGEX, version):
            self._extension = "alpha"
            version = version[: version.rfind(_EXTENSION_SEPARATOR)]
            parts = version.split(_VERSION_SPLITTER)
            self.set_major(0)
            self.set_minor(int(parts[1]))
            self.set_micro(int(parts[2]))
        elif re.match(_VALID_BETA_REGEX, version):
            self._extension = "beta"
            version = version[: version.rfind(_EXTENSION_SEPARATOR)]
            parts = version.split(_VERSION_SPLITTER)
            self.set_major(0)
            self.set_minor(int(parts[1]))
            self.set_micro(int(parts[2]))
        elif re.match(_VALID_RELEASE_REGEX, version):
            self._extension = None
            parts = version.split(_VERSION_SPLITTER)
            self.set_major(int(parts[0]))
            self.set_minor(int(parts[1]))
            self.set_micro(int(parts[2]))
        else:
            raise ValueError("Invalid input!")

    def get_major(self) -> int:
        return self._major

    def set_major(self, major: int) -> None:
        self._major = major

    def get_minor(self) -> int:
        return self._minor

    def set_minor(self, minor: int) -> None:
        self._minor = minor

    def get_micro(self) -> int:
        return self._micro

    def set_micro(self, micro: int) -> None:
        self._micro = micro

    def is_pre_alpha(self) -> bool:
        return self.get_major() == 0 and self.get_minor() == 0 and self._extension is None

    def is_alpha(self) -> bool:
        if self._extension is None:
            return False
        return self._extension == "alpha"

    def is_beta(self) -> bool:
        if self._extension is None:
            return False
        return self._extension == "beta"

    def is_release(self) -> bool:
        return not (self.is_pre_alpha() or self.is_alpha() or self.is_beta()) and self._extension is None

    @staticmethod
    def compare_versions(older: "SemVerUtil", newer: "SemVerUtil") -> int:
        """Return 0 if equal, 1 if ``newer`` is ahead of ``older``, -1 otherwise."""
        if str(older) == str(newer):
            return 0
        if older.is_pre_alpha() and not newer.is_pre_alpha():
            return 1
        if older.is_alpha() and (newer.is_beta() or newer.is_release()):
            return 1
        if older.is_beta() and newer.is_release():
            return 1
        all_flags = (
            older.is_pre_alpha() == newer.is_pre_alpha()
            or older.is_alpha() == newer.is_alpha()
            or older.is_beta() == newer.is_beta()
            or older.is_release() == newer.is_release()
        )
        if all_flags:
            if older.get_major() < newer.get_major():
                return 1
            if older.get_major() == newer.get_major() and older.get_minor() < newer.get_minor():
                return 1
            if (
                older.get_major() == newer.get_major()
                and older.get_minor() == newer.get_minor()
                and older.get_micro() < newer.get_micro()
            ):
                return 1
        return -1

    def is_backwards_compatible_with(self, other: "SemVerUtil") -> bool:
        # TODO: Two versions are backwards compatible if they are equal or if the major and minor versions are equal
        return False

    def __str__(self) -> str:
        extension = "" if self._extension is None else "-" + self._extension
        return f"{self.get_major()}.{self.get_minor()}.{self.get_micro()}{extension}"
